### DISCUSSION TITLE
Use json as embeddedLanguage in custom grammar injection

### DISCUSSION THREAD

**Discussion by chrisrhymes** [microsoft/vscode-discussions#251] [Extension Development QnA] [answered]
Hi, I'm writing a VS Code extension that will help with WordPress Gutenburg block development. Currently the blocks appear as a HTML comment and is all greyed out, so I want to be able to add syntax highlighting to the comment to make them easier to read and add json support. 

This is an example tag you might find. 

```html
<!-- wp:column {"align":"wide"} -->
```

Here is my repo: https://github.com/chrisrhymes/gutenburg-comment-highlight

## Embedded Language

The extension currently matches the string between the curly braces and identifies it as `source.json` but the language is still set to HTML, even though I have set `source.json` to be `json` in the embeddedLanguages in the package.json.

```json
"embeddedLanguages": {
    "source.json": "json"
},
```
 
<img width="809" alt="Screenshot 2022-11-17 at 10 04 18" src="https://user-images.githubusercontent.com/4160546/202419057-9783c1a6-dd5b-4446-8f63-60c738ed6eda.png">

If I set the match to be `source.js` and update the embeddedLanguages to be javascript it correctly detects the language as javascript, but doesn't work for json. 

```json
"embeddedLanguages": {
    "source.js": "javascript"
},
```

<img width="808" alt="Screenshot 2022-11-17 at 10 03 37" src="https://user-images.githubusercontent.com/4160546/202420187-537b801b-c080-4106-afa1-93f54939fdcf.png">

## Token Types

The other issue I have is that it is still setting the standard token type to Comment instead of Other, so no syntax highlighting is working. I have added the `"source.json": "other"` to tokenTypes in the package.json but this seems to be ignored. 

```json
"tokenTypes": {
    "source.json": "other"
}
```

Thanks!

**Comment by chrisrhymes** [comment, marked as answer]
Looks like I needed to include source.json in the patterns in the grammar and change the name to meta.embedded.block.json to get it to set the language as json and the token type to other. 

```json
"gutenburg-json": {
      "match": "{([^}]+)}",
      "name": "meta.embedded.block.json",
      "patterns": [
        {
          "include": "source.json"
        }
      ]
    }
```

I still need to do some work on the regex as it currently stops the match after the first `}` if there is a nested object.

**Reply by isidorn** [reply]
fyi @aeschli

**Comment by JustinGrote** [comment]
This definitely needs to be noted in the  grammar embedded authoring docs, as it drove me crazy why it wasn't working till I found this.

**Reply by JustinGrote** [reply]
Also it appears that even if you use "match", highlighting doesn't occur unless you explicitly use "begin" and "end", instead even tho the language appears to detect as JSON just fine.
![image](https://github.com/microsoft/vscode-discussions/assets/15258962/e4963e6b-93f8-4442-b382-4e04d1c14392)

**Comment by gjsjohnmurray** [comment]
> This definitely needs to be noted in the grammar embedded authoring docs, as it drove me crazy why it wasn't working till I found this.

@JustinGrote I have always found the documentation team very receptive to PRs, and they make it easy by offering an "Edit" button at the top right of most pages, e.g. https://code.visualstudio.com/api/language-extensions/embedded-languages
